auv_navigation/auv_navigation/scripts/follow_path_action_server.py

```diff
@@ -0,0 +1,233 @@
+                    )
+                    if trans_error is None or rot_error is None: 
+                        rospy.logwarn(f"Failed to get transform between {self.source_frame} and {self.carrot_frame}")
+                        self.control_rate.sleep() 
+                        continue
+                    
+                    if self.frame_controller.is_aligned(self.position_threshold, self.angle_threshold,
+                                                        current_pose=current_pose, path=path, trans_error=None, rot_error=None):  
+                        rospy.loginfo("Goal reached via path following!")
+                        return True
+                    
+                    cmd_vel = self.frame_controller.compute_cmd_vel(trans_error, rot_error)
+                    self.cmd_vel_pub.publish(cmd_vel)
+                    
+                    feedback.distance_to_goal = np.sqrt(
+                        (path.poses[-1].pose.position.x - current_pose.pose.position.x) ** 2 +
+                        (path.poses[-1].pose.position.y - current_pose.pose.position.y) ** 2 +
+                        (path.poses[-1].pose.position.z - current_pose.pose.position.z) ** 2
+                    )
+                    self.server.publish_feedback(feedback)
+                        
+                # 2. If path is not valid, do direct alignment
+                elif path is None: 
+                    
+                    trans_error, rot_error = self.frame_controller.get_error(
+                        self.tf_buffer,
+                        self.source_frame,
+                        target_frame,
+                        angle_offset,
+                        keep_orientation
+                    )
+                    if trans_error is None or rot_error is None: 
+                        rospy.logwarn(f"Failed to get transform between {self.source_frame} and {target_frame}")
+                        self.control_rate.sleep() 
+                        continue
+                    
+                    if self.frame_controller.is_aligned(self.position_threshold, self.angle_threshold,
+                                                        current_pose=None, path=None, trans_error=trans_error, rot_error=rot_error):                                      
+                        rospy.loginfo("Goal reached via direct alignment!") 
+                        return True
+                    
+                    cmd_vel = self.frame_controller.compute_cmd_vel(trans_error, rot_error)
+                    self.cmd_vel_pub.publish(cmd_vel)
+
+                self.control_rate.sleep()
+
+            return False # If exited the loop, success is False
+        except Exception as e:
+            rospy.logerr(f"Error during alignment: {e}") 
+            return False
+
+    def execute(self, goal):
+        rospy.loginfo("FollowPathActionServer: Received a new FollowPath goal.")
+        
+        if goal.path and goal.path.poses:
```

Comment:
this approach is rather confusing. the action or server or the topic fields should not be populated selectively. if a secondary type of input is required, then you should have secondary topic (or action namespace) with it's own action message types.


 so either poses are provided 
 or 2 frames. I'm also in favor of separating that to another node which creates poses from the given 2 frames. the output of that node is connected to this node. etc etc